### ISSUE TITLE
Profile and optimise PDE Jacobian computation (primary training bottleneck)

### ISSUE THREAD

**Issue by zeinali72** [zeinali72/swe-pinn#105]
## Context

The PDE loss computation in \`src/losses/pde.py\` is almost certainly the **single largest bottleneck** in training. Every gradient step computes:

1. **Forward Jacobian**: \`jax.vmap(jax.jacfwd(U_fn))(pde_batch)\` — full 3x3 Jacobian (3 outputs h,hu,hv × 3 inputs x,y,t) for every PDE point
2. **Backward pass through Jacobian**: \`jax.grad(loss_fn)\` backpropagates through the Jacobian computation — effectively computing **second-order derivatives** (Hessian-vector products)

This double autodiff happens on every batch, every epoch.

## Related Issues

- **Depends on**: #104 (profiling infrastructure — need timing tools first)
- **Parallel-safe** with: #105 (memory profiling), #106 (data transfer profiling)
- **Related**: #60 (batch cycling — affects how often PDE is computed)

## Current Code

\`\`\`python
# src/losses/pde.py, lines 13-58
def compute_pde_loss(model, params, pde_batch, config, pde_mask=None):
    def U_fn(pts):
        return model.apply({'params': params['params']}, pts, train=False)

    jac_U = jax.vmap(jax.jacfwd(U_fn))(pde_batch)   # ← BOTTLENECK
    dU_dx = jac_U[..., 0]
    dU_dy = jac_U[..., 1]
    dU_dt = jac_U[..., 2]
    # ... SWE residual computation using Jacobians
\`\`\`

And in the boundary losses:
\`\`\`python
# src/losses/boundary.py, lines 57-65
def loss_boundary_neumann_outflow_x(model, params, batch):
    def U_fn(pts):
        return model.apply({'params': params['params']}, pts, train=False)
    jac_U = jax.vmap(jax.jacfwd(U_fn))(batch)        # ← SECONDARY BOTTLENECK
    dU_dx = jac_U[..., 0]
\`\`\`

### 1. Measure Jacobian wall-clock time

\`\`\`python
# Isolate and time the Jacobian computation
pde_batch = sample_domain(key, n_points, ...)
jac_fn = jax.jit(jax.vmap(jax.jacfwd(U_fn)))

# Warmup (JIT compile)
_ = jac_fn(pde_batch[:1])
jax.block_until_ready(_)

# Benchmark
import timeit
for n in [100, 500, 1000, 5000, 10000]:
    batch = pde_batch[:n]
    t = timeit.timeit(lambda: jax.block_until_ready(jac_fn(batch)), number=10) / 10
    print(f"  n={n:>6}: {t*1000:.1f} ms ({t/n*1e6:.1f} us/point)")
\`\`\`

### 2. Measure full gradient step time with breakdown

Profile \`train_step_jitted\` and break down time by loss component:
- PDE Jacobian computation
- SWE residual calculation
- IC loss (forward pass only — no Jacobian)
- BC losses (Dirichlet: forward only; Neumann: has Jacobian)
- Backward pass (grad through all of the above)
- Optimizer update

### 3. Compare \`jacfwd\` vs \`jacrev\` for this model

\`\`\`python
# jacfwd: O(n_inputs) forward passes → 3 forward passes for (x,y,t)
# jacrev: O(n_outputs) backward passes → 3 backward passes for (h,hu,hv)
# Since n_inputs == n_outputs == 3, they should be similar, but memory differs

jac_fwd = jax.jit(jax.vmap(jax.jacfwd(U_fn)))
jac_rev = jax.jit(jax.vmap(jax.jacrev(U_fn)))

# Compare time and memory for both
\`\`\`

### 4. Benchmark per architecture

The Jacobian cost depends heavily on model architecture:
- **MLP**: Straightforward — Jacobian of dense layers
- **FourierPINN**: Fourier feature layer adds computation to Jacobian
- **DGM**: Gated sub-networks in every layer — most expensive Jacobian

Profile all three at the same width/depth to compare.

## Potential Optimisations to Evaluate

1. **Custom JVP rules** — if the SWE only needs specific Jacobian columns, avoid computing the full 3x3
2. **Chunk the vmap** — \`jax.vmap\` over 10k points may exhaust GPU memory; chunking with \`jax.lax.map\` could help
3. **Mixed precision** — compute Jacobian in float16, accumulate in float32
4. **Reduce PDE batch size** — if Jacobian is the bottleneck, fewer PDE points per batch may improve wall-clock time per epoch
5. **Jacobian caching** — if learning rate is small, Jacobian changes slowly; reuse across consecutive steps (aggressive, needs validation)

## Files to Profile

- \`src/losses/pde.py:compute_pde_loss\` — primary target
- \`src/losses/boundary.py:loss_boundary_neumann_outflow_x\` — secondary target
- \`src/training/step.py:train_step\` — full step timing

## Acceptance Criteria

- [ ] Wall-clock timing for Jacobian computation at various batch sizes
- [ ] Breakdown of train_step time by loss component
- [ ] \`jacfwd\` vs \`jacrev\` comparison
- [ ] Per-architecture benchmark (MLP vs FourierPINN vs DGM)
- [ ] Documented results in \`profiling/pde_jacobian_benchmark.md\` or similar
